### ISSUE TITLE
Develop data backup and recovery plan

### ISSUE THREAD

**Issue by jadudm** [GSA-TTS/FAC#2145]
# At a glance

[comment]: # "Begin with a short summary so intent can be understood at a glance."

[comment]: # "In order to: some objective or value to be achieved"
[comment]: # "as a: stakeholder"
[comment]: # "I want: some new feature"

**In order to** develop tooling and standard operating procedures around data backups
**as a** product owner
**I want** a thought-out plan for what we will develop and do.

# Acceptance Criteria

We use [DRY](https://docs.behat.org/en/latest/user_guide/writing_scenarios.html#backgrounds) [behavior-driven development](https://en.wikipedia.org/wiki/Behavior-driven_development#Behavioral_specifications) wherever possible.

[comment]: # "ACs should be clearly demoable/verifiable whenever possible."
[comment]: # "Given: the initial context at the beginning of the scenario"
[comment]: # "when: the event that triggers the scenario"
[comment]: # "then: the expected outcome(s)"
[comment]: # "Repeat scenarios as needed, or repeat behaviors and lists within a scenario as needed."

[comment]: # "The scenario should be a short, plain language description."
[comment]: # "Feeling repetative? Apply the DRY (Don't Repeat Yourself) principle!"

The working scenario for this planning is the following:

1. We have developed and tested a feature in `dev`
2. ... deployed it to `preview` to test with users
3. ... deployed successfully with regression tests to `staging`
4. ... and pushed to `prod`, where our code thrashes our production database, leaving us in an inconsistent or uncertain state.

The question becomes: how quickly can we restore to a known good state, and how close to **now** can we reach back? 

In this way, we're looking for a process that we can run on every deploy to `prod`, that we can test on every deploy to `staging`, and that we can (if we choose) also use for additional data snapshots throughout the day.

We have implemented some initial tooling that will backup our databases to an S3 bucket. This is excellent. Before we proceed further, we now want to get multiple members of the team up to speed as to 1) what has been done and 2) why. 

Then, before diving in and implementing more code, we want to identify what we want to achieve, how we will do the work in service to those goals, and develop stories and tickets to track the work. 

So, in order to move from a single script that dumps our database as a result of a `cf task` call to a reliable and robust process for dumping, verifying, and restoring our production data...

[comment]: # "Each task should be a verifiable outcome"
```[tasklist]
### then...
- [x] share with more members of the team what was done and why
- [x] describe what data will be backed up in this increment of work
- [ ] develop a working set of criteria for what we should consider as "done" in terms of data backup and restore
- [ ] provide a high-level sketch of what the tool or tools involved will need to do
- [ ] provide a look-forward to how this process will support/need to change in light of historic data migration
```
List of tasks to meet the objectives listed in the GDoc:
```[tasklist]
- [ ] https://github.com/GSA-TTS/FAC/issues/2211
- [ ] Backup data in prod every night
- [ ] Create a manually triggered workflow for FAC Admin to pick a prod backup and restore to prod
- [ ] Create a cleansed prod backup that can be used in dev, preview and local envirments
- [ ] Create a manually triggered workflow for FAC Admin to pick a cleansed backup and restore to dev or preview
- [ ] Create scripts to fetch a cleansed backup and load into a local environment
```
### Shepherd

[comment]: # "@ mention shepherds as we move across the board."

* Engineering shepherd: @JeanMarie-TTS 

# Background

[comment]: # "Any helpful contextual notes or links to artifacts/evidence, if needed"

Some initial thoughts:

1. I believe this planning should initially be constrained to our **live production data**. It is reasonable if the first increment focuses on the *intake* table(s). 
6.  A good solution will easily extend to cover the *dissemination* tables and any tables that are required to support operation. (This does *not* include our pilot historical data, for instance. That is/can be restored via other means.)
7. We should be able to validate the integrity of our export. Did we export all of the data? How would we know? Did we export it with fidelity, or did we introduce errors? Again, how would we know? (Should we hash every row and compare to living data as we perform our backup operations? Should we hash and compare as we restore?)
8. We should be able to validate the integrity of our restore process. Did all the data come back that we expected to come back?
9. We should be able to run the restore process against a live, production environment, and not endanger the living data. That is, a restore run against `prod` should not damage the data in `prod`. 
10. We should not tackle larger questions in this work: for example, how to back up S3 buckets.
11. From a security perspective, do we have a way of knowing that data we backed up at a given moment is still the same when we restore it? (This may be a step too far at this point, but we should be thinking about it.)

I am looking for a document that can be presented to the team for consideration. I'm happy to provide feedback as the work progresses. The document needs to be functional, and it should not leave large gaps to the imagination. It will serve to develop the stor(y|ies) for the work that will follow.

#2119 was a triage/placeholder ticket for this story. Some conversation started there, but should continue here and in any planning documents developed.

# Security Considerations

Required per [CM-4](https://nvd.nist.gov/800-53/Rev4/control/CM-4).

[comment]: # "Our SSP says 'The team ensures security implications are considered as part of the agile requirements refinement process by including a section in the issue template used as a basis for new work.'"
[comment]: # "Please do not remove this section without care."
[comment]: # "Note any security concerns that might be implicated in the change. 'None' is OK, but we must be explicit here."

We need to think about whether our backups are always attached, as their integrity does matter. We can achieve some of this by limiting the exposure to the backup buckets.

---

<details>
  <summary>Process checklist</summary>
  
# Sketch

[comment]: # "Notes or a checklist reflecting our understanding of the selected approach"

- [ ] Design designs all the things
- [ ] Engineering engineers all the things

# Definition of Done

### If not likely to be important in the next quarter...
- [ ] Archived from the board

### Otherwise...

- [ ] Has a clear story statement
- [ ] Design or Engineering accepts that it belongs in their respective backlog

## Design Backlog

- [ ] Has clearly stated/testable acceptance criteria
- [ ] Meets the design Definition of Ready [citation needed]
- [ ] A design shepherd has been identified

## Design In Progress

- [ ] Meets the design Definition of Done [citation needed]

## Design Review Needed

- [ ] Necessary outside review/sign-off was provided

## Design Done

- [ ] Presented in a sprint review
- [ ] Includes screenshots or references to artifacts

### If no engineering is necessary
- [ ] Tagged with the sprint where it was finished
- [ ] Archived

## Engineering Backlog

- [ ] Has clearly stated/testable acceptance criteria
- [ ] Has a sketch or list of tasks
- [ ] Can reasonably be done in a few days (otherwise, split this up!)

## Engineering Available

- [ ] There's capacity in the `In Progress` column
- [ ] An engineering shepherd has been identified

## Engineering In Progress

- [ ] Meets acceptance criteria
- [ ] Meets [QASP conditions](https://derisking-guide.18f.gov/qasp/)

### If there's UI...
- [ ] Screen reader - Listen to the experience with a screen reader extension, ensure the information presented in order
- [ ] Keyboard navigation - Run through acceptance criteria with keyboard tabs, ensure it works. 
- [ ] Text scaling - Adjust viewport to 1280 pixels wide and zoom to 200%, ensure everything renders as expected. Document 400% zoom issues with USWDS if appropriate.

## Engineering Blocked

- [ ] Blocker removed/resolved

## Engineering Review Needed

- [ ] Outside review/sign-off was provided

## Engineering Done

- [ ] Presented in a sprint review
- [ ] Includes screenshots or references to artifacts
- [ ] Tagged with the sprint where it was finished
- [ ] Archived

</details> 
```[tasklist]
### Tasks
```

**Comment by JeanMarie-TTS** [comment]
@asteel-gsa , @mogul  Started a working GDoc to discuss and decide on our approach - see https://docs.google.com/document/d/13QBxMzjpaBWAtG0pGnOAwc8Ct9TAh0-qnkKe6Bqrkg4/edit. We will create an ADR and engineering stories/tasks once we finalize the design.
Please edit and comment.

**Comment by jadudm** [comment]
As you requested in Slack, you wanted clarification regarding scope:

The scope of this ticket is our data in our databases, not media files in our S3 buckets. That can, and should be, a separate piece of work.

If you have questions, feel free to ask. As the story and tasks clearly imply:


> ... in order to move from a single script that dumps our database as a result of a cf task call to a reliable and robust process for dumping, verifying, and restoring our production data...

The Background section also makes clear that this is about the database data, not the media files.

**Comment by JeanMarie-TTS** [comment]
List of tasks to meet the objectives listed in the GDoc:

- [ ] Backup data in prod whenever there is a deployment
- [ ] Backup data in prod every night
- [ ] Restire data from prod to staging every night
- [ ] Create a manually triggered workflow for FAC Admin to pick a prod backup and restore to prod
- [ ] Create a cleansed prod backup that can be used in dev, preview and local envirments
- [ ] Create a manually triggered workflow for FAC Admin to pick a cleansed backup and restore to dev or preview
- [ ] Create scripts to fetch a cleansed backup and load into a local environment

**Comment by JeanMarie-TTS** [comment]
@jadudm 
Please take a look at the tasks list above and assign MSCW riorities. I will convert them to tickets prioritizing the MS ones.

Please note the following:
- Our backups will backup all the data in the postgres instance. Thus there is no distinction between audit, dissemination, etc.
- When we restore to prod, we will have to take the app offline. The restore **replaces** the current data with the data from the backup.
- Backing up and restoring django media is just one additional command, and we have this working. Decoupling media will require us to figure out how to ensure integrity between media and db data, which is a harder problem.
- Not sure what we want to do regarding #7 in your list. If we are concerned with the integrity of what's in our S3 backup bucket, we should be concerned with the integrity of our data in other S3 buckets and perhaps even in the database. I think we should just rely on cloud.gov to ensure this.

**Comment by jadudm** [comment]
Ha! GH linked to ticket seven... that confused me for a second.

I think what I mean by integrity (in particular, for the tabular data, to start) is this:

I dump a database.

Is what is in the dump the same as what was in the database? Should we, after a dump, check (at the least) that the same number of rows are present in the dump as there are in the live DB at time of dump? should we checksum the data?

Later, I do a reload.

Is the data that is restored the same as the data that was in the dump?

In short

dump -> verify

load -> verify

Perhaps that's a given in what you've described. But the mantra I've heard before is "a backup that is not tested is not a backup." So, how are we going to test our backups? I think the ADR piece can be ignored for the moment; we'll document what we do when we're done. 

Your point about the media makes sense.

**Comment by jadudm** [comment]
The epic #2211 picked up the work of documentation and storying, and provided more refined tickets for all the tasking. Closing this initial story in lieu of that epic and better tickets. 

That epic was created based on

https://github.com/GSA-TTS/FAC/issues/2145#issuecomment-1729200901

four days ago, and is a better place to track the work.
